### ISSUE TITLE
Printing Notation regressed compared to 8.7

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#6764]
#### Version

git master

#### Operating system
Linux

#### Description of the problem

In the `gen_proofmode` branch of https://gitlab.mpi-sws.org/FP/iris-coq/, printing notations regressed compared to Coq 8.7. This is visible for example in `theories/bi/update.v`: What used to be printed as `(|==> P) ⊢ P` is now printed as `(sbi_bi PROP).(bi_entails) (|==> P)%I P`.

The relevant notation `⊢` is defined in `theories/bi/interface.v`:
```
Notation "P ⊢ Q" := (bi_entails P%I Q%I) : stdpp_scope.
```
`bi_entails` is a primitive projection, so #6651 could be related. Or maybe this is caused by one of the recent scope handling changes?

**Comment by RalfJung** [comment]
Small example:
```coq
Set Primitive Projections.

Record foo := Foo { foo_n : nat }.

Notation "'■' x" := (foo_n x%type) (at level 50).

Goal forall (f:foo), ■ f = ■ f.
```
With 8.7, this prints what I typed (`■ f = ■ f`), with master it prints `f.(foo_n) = f.(foo_n)`

**Comment by JasonGross** [comment]
This is almost certainly #6651, because the printing worked fine on the only-slightly-outdated version of master I was previously using (47e43e229ab02a4dedc2405fed3960a4bf476b58)

**Comment by RalfJung** [comment]
Ah, you're right. I thought the `%type` was the problem, but actually the problem persists without it.

**Comment by ppedrot** [comment]
Yes, the problem is that notation matching is completely unaware of primitive projections. The input term to be matched is pretyped to a primitive projection and then back to a `GProj`, but the notation constr doesn't know anything about that as it is a `NApp` node. Obviously matching then fails because the two terms are unrelated.

Either we need to handle specially primitive projections in notation matching, or we need to also generate a notation rule for applied primitive projections. I don't know which one is better.

**Comment by ppedrot** [comment]
I have a potential patch at ppedrot/coq@189528e253df092ac5d542cb945139649e5edce0 but it breaks due to pretyping being stupid in the handling of `GProj`, as hinted by the following comment.
```
    (* TODO: once GProj is used as an input syntax, use bidirectional typing here *)
```
On @RalfJung's example I get a pretype error saying that the hole cannot be inferred.

**Comment by maximedenes** [comment]
I didn't follow how the input `GProj` is produced, but in any case, it shouldn't be too hard to implement the bidirectional typing rule.

**Comment by maximedenes** [comment]
But I'm a bit worried that some other parts of the code could distinguish the `GApp` nodes from `GProj`, which is why I tried to avoid introducing them (except in the printing, where I forgot to handle notations for such projections).

**Comment by ppedrot** [comment]
Ultimately, I think my local solution is the best as it tries not to introduce ad-hoc code. Whether this is sane or not w.r.t. bug preservation is another story.

**Comment by herbelin** [comment]
Good to see this work in making primitive projections better and better! I have a few comments.

- As pointed out by @mattam82 in this [comment](https://github.com/coq/coq/pull/6651#pullrequestreview-94637300), there should eventually be a `ProjRef` global reference so that we don't need to call functions such as `Environ.is_projection`. Has someone already plans to go in this direction?
- We need a rationale to explain whether `t.(p)` is intended to mean a (negative) primitive projection `GProj` or to also mean a compact representation of the (positive) defined named projection `GProj false`? I believe that it would be less a source of confusion to mean only the first	situation, but, by compatibility, we may also want the second.
- In my view, I don't think that `NProj` should come with a Boolean. It should only correspond to `GProj false` (in the positive case) and to the only `GProj` which matters in the negative case. Even for `GProj`, it is not clear to me that it should come with an `unfolded` flag (compared to elaborating a `Proj true` only at pretyping time). This is consistent with having `CProj` bound to `GProj false` and no name-based syntax for `GProj true` (i.e. the syntax for `GProj true` should be a special form of `match` in the positive case and being irrelevant in the negative case).

Again, I have no idea about whether we have a common view on the meaning to give to the `unfolded` flag and on the "negative" vs "positive" distinction, so please comment to tell your view. It may also be the case that I oversought some discussions on this topic, so please tell me if this is the case.

PS: In line 366 [here](https://github.com/ppedrot/coq/commit/189528e253df092ac5d542cb945139649e5edce0#diff-9d38ecdafdebc0e7282b1341b183e1fbR366), couldn't there be the case of a projection with a functional type and more arguments coming after the main argument?

**Comment by ppedrot** [comment]
> I have no idea about whether we have a common view on the meaning to give to the unfolded flag 

I'd be glad if @mattam82 gave a comprehensive description of the uses of that flag, actually.

> PS: In line 366 here, couldn't there be the case of a projection with a functional type and more arguments coming after the main argument?

Yes, I thought about it but I wondered whether one could get around that by using explicit parentheses? I don't remember if we collapse `GApp` nodes during internalization.

**Comment by herbelin** [comment]
> Yes, I thought about it but I wondered whether one could get around that by using explicit parentheses? I don't remember if we collapse GApp nodes during internalization.

Actually, I'm lost. We have two representations of `foo_n x` in `glob_constr`? Is it because we want to preserve a different printing for them?

Actually, I'm not sure that I understand very clearly how to interpret a non-fully-applied projection, nor if we should keep an explicit representation for projections with all its parameters which is distinct from the optimized form.

For instance, let's consider the following cases
```
Set Primitive Projections.
Record foo := Foo { foo_n : nat }.
Notation "'■' x" := ((x%type).(foo_n)) (at level 50).
Check forall (f:foo), ■ f = ■ f.
(* Should use the notation by principle of reversibility *)
Check forall (f:foo), foo_n f = foo_n f.
(* Should it use the notation, i.e. is it a different expression or not? *)
Notation "# x" := (foo_n (x%type) ) (at level 50).
Check forall (f:foo), # f = # f.
(* Should use the "#" notation by principle of reversibility *)
Check forall (f:foo), f.(foo_n) = f.(foo_n).
(* Should it use the "#" notation or the "■" notation, i.e. is it a different expression or not? *)
```
(For the record, in today's master, it does not use the notation in either cases.)

Otherwise said, shouldn't there be a difference between `foo_n x` (meaning a `Const` with all its parameters explicit) and `x.(foo_n)` (meaning a `Proj false` with its parameters absent)? By doing so,`foo_n` alone would make sense as the stand-alone constant, `foo_n p` would make sense as this constant applied to parameters (at least in printing), while in parsing, we would have the choice. To take benefit of the compact representation of `foo_n p`, we could directly reinterpret it (and thus print it) as `p.(foo_n)`? Or, alternatively, to be consistent with the internal representation, we interpret it as `foo_n all_its_params p` (but loosing the compactness of the representation).

**Comment by ppedrot** [comment]
@herbelin I think that the compatibility layer is the source of all this fuss. There shouldn't be a confusion between the primitive projection and the corresponding constant. I understand this would be a problem for porting code to primitive projections, but at the same time in the long run it'd save us a lot of ultimately nonsensical questions.

The current situation regarding `glob_constr` is that:
* When the user writes a term standing for a projection, whether using `x.(foo)` or `foo x`, this is always desugared to a `GApp` by internalization. It is the rôle of the pretyper to trigger ad-hoc code recognizing that a fully applied constant is interpreted as a `mkProj`.
* Nonetheless, since @maximedenes's patch, the detyper produces `GProj` nodes when handling primitive projections. So there is an asymmetry here, which is what the notation bug uncovers.

**Comment by RalfJung** [comment]
I think it would be nice to have an option for `proj t`, when fully applied, to be turned into `t.(proj)`. Without such an option, we would suffer a severe performance regression as we'd not actually use the primitive projections most of the time. Also, I cannot imagine a situation where one writes a fully applied projection and does not want the compact (parameter-free) representation.

Another open question is: Is there a way to make `t.(proj)` fail if this is not a primitive projection? IOW, it would be nice to have a guarantee that, when writing `t.(proj)`, we get the compact representation.

**Comment by herbelin** [comment]
@RalfJung 

> Is there a way to make t.(proj) fail if this is not a primitive projection? 

BTW, related question: would you like the notation `{|... ; proj := t; ...|}` to fail if not in a type equipped with primitive projections?

Or is your question strictly only about the compactness of the representation of projections and not about the introduction/destruction/reduction/eta rules which are associated to the type, i.e. not about whether these projections are primitive in the logical meaning of "primitive" (i.e. with surjective pairing as eta, without a proper `match`, without `unfold` of a projection being allowed) but really about whether they are represented compactly, independently of whether it is an optiimization of the good old existing form of (co)inductive types (with `match` and `unfold`-able projections) or a new primitive form of type built from projections (typically in the Abel-Pietka-Sentzer style)? (Hoping that my question makes clear enough sense.)

**Comment by herbelin** [comment]
Actually, do we really need to keep types without a compact representation of projections?

Is it correct that the only incompatibility (in the "compatibility" mode) is with tactics counting or unfolding subterms occurring in the parameters of a projection? Or maybe also sometimes in how types are inferred when the arguments of a projection are implicit? So, in a context where efficiency matters for scaling, the `Compatibiility` `Primitive Projections` mode should be on by default, no? And deactivated only when there are too much worries with the irreducible (little?) incompatibilities due to having the parameters of the projection canonically inferred rather than explicitly given?

**Comment by herbelin** [comment]
Let me rephrase things once more differently.

There are basically three independent design choices to do for types equipped with projections:
1. whether they are defined from projections or defined from `match` and constructors
2. whether they represent projections in a compact way or not
3. whether the projections are represented with the `t.(proj)` notation or not (and subsidiary, whether tuples are represented with the `{| ... |}` notation or not)

Currently, 3. is independent of 1. and 2., and this is precisely the point of @RalfJung's question and of @maximedenes's #6651 to tie 3. to either 1. or to 2.

Currently, demanding a compact representation in 2. is not the default and this is precisely the point of @JasonGross's #6609 to always adopt a compact representation (a priori independently of the degrees of freedom 1. and 3.).

**Comment by RalfJung** [comment]
> would you like the notation {|... ; proj := t; ...|} to fail if not in a type equipped with primitive projections?

I have not thought about that. I was coming mostly from the performance stand-point -- the reason we use primitive projections is because it makes things measurably faster; if we can't even see in the surface syntax whether we are actually applying that optimization, that would be problematic and it would make performance issues harder to debug.

**Comment by maximedenes** [comment]
I didn't manage to fix this in time for the beta, sorry about that.

If we don't come up with a better with, one easy thing we could do for the final version is to unplug the construction of `GProj`, so as to undo the effects of my PR while keeping the infrastructure.

**Comment by Zimmi48** [comment]
This was fixed by the revert of #6651, right?

**Comment by RalfJung** [comment]
Yes seems to be fixed in 8.8.dev :-)

**Comment by maximedenes** [comment]
Yes, but the bug is still present in `master`, so let's keep the issue open.

**Comment by RalfJung** [comment]
Okay, sorry.

**Comment by RalfJung** [comment]
@maximedenes 8.9 is coming closer and this is still open. Right now, 8.9 is not usable for Iris. Are there any plans to address this?

**Comment by maximedenes** [comment]
Let's allocate some time to sort this out, and revert if we can't. @ppedrot do you have some time tomorrow to discuss the appropriate course of action?

**Comment by ppedrot** [comment]
That should be possible indeed.

**Comment by RalfJung** [comment]
Did that meeting come to any conclusion?

**Comment by maximedenes** [comment]
> Did that meeting come to any conclusion?

Yes, I'll try to submit a patch tomorrow.

**Comment by silene** [comment]
If I understand correctly, 8.8 is already broken, so 8.9 is not a regression. In other words, this is not a regression, as far as I am concerned. (That said, this is certainly an ugly bug, so I will backport the fix if nothing goes wrong with it.)

**Comment by RalfJung** [comment]
8.8 works fine (it is what I usually use for Iris development), the patch introducing the regression got reverted in that branch. So this is definitely a regression from 8.8 to 8.9.

**Comment by maximedenes** [comment]
Ok, the fix is not easy, and it won't be tested in time. So I'll prepare a revert.

**Comment by maximedenes** [comment]
@RalfJung should be fixed by #8456

Sorry for the delay, I really hoped we could do better.

**Comment by gares** [comment]
> Ok, the fix is not easy

It may help to write here why, so that we don't forget. How should the reverted patch be improved?

**Comment by maximedenes** [comment]
> It may help to write here why, so that we don't forget. How should the reverted patch be improved?

It should be improved by dropping the special status of compatibility constants and implementing bidirectionnal pretyping of `GProj`.

**Comment by maximedenes** [comment]
@RalfJung can you confirm that the Iris problems are solved?

**Comment by ejgallego** [comment]
Indeed @RalfJung it would be good to know if additional problems with IRIS do exist in 8.9 and `master`.

**Comment by RalfJung** [comment]
As confirmed on gitter, this fixes the biggest problem. Our test suite still failed because backtraces for `first` and `last` tacticals got "improved" but [we don't want these backtraces](https://github.com/coq/coq/issues/7769), I have fixed that since then and now Coq master passes our test suite just fine.

Is there an issue tracking getting this backported into 8.9?

**Comment by ejgallego** [comment]
Great @RalfJung , it should be scheduled for backporting due to the milestone.

**Comment by Zimmi48** [comment]
This was an API breaking change so this is why it takes longer to backport.
Cf. https://github.com/coq/coq/pull/8456#issuecomment-427272829
cc @gares

**Comment by RalfJung** [comment]
I am going to reopen this issue until a fix lands in 8.9.

@gares what's the status of the backport?

**Comment by gares** [comment]
I made an overlay for elpi I think I made a pr... Dunno if someone wrote the overlay file

**Comment by gares** [comment]
https://github.com/LPCIC/coq-elpi/tree/overlay/8456?files=1 sorry I made a branch, dunno for the overlay file

**Comment by RalfJung** [comment]
@Zimmi48 @ejgallego is that enough to get a backport started?

**Comment by Zimmi48** [comment]
@gares Currently CI for v8.9 is testing commit gares/coq-elpi@12efa57c37145dcfe331e6350b5d5cc145b13807, i.e. before #8552 (using elpi 1.1.0). Creating a v8.9 branch starting from this commit and backporting the fix to this branch (no need to go through a PR), would allow @silene to backport #8456 without backporting #8552.

**Comment by gares** [comment]
I made a coq-v8.9 branch on master, since master compiles fine with 8.9, and cherry picked there the commit. So the coq-v8.9 branch should compile fine when CProj is removed.

I don't get exactly the before #8852 part. Are you saying that the docker image for v8.9 does not contain elpi 1.1.0 ?

**Comment by Zimmi48** [comment]
>Are you saying that the docker image for v8.9 does not contain elpi 1.1.0 ?

Exactly.

**Comment by ejgallego** [comment]
> Exactly.

The elpi update is supposed to be in the backport queue, right?

**Comment by Zimmi48** [comment]
It is still in the backport queue indeed. But @silene was hoping he could avoid backporting it. That's the sense of his comment and mine here: https://github.com/coq/coq/pull/8456#issuecomment-427272829

**Comment by gares** [comment]
I don't how much relevant elpi is, but I've just release the first version of coq-elpi and I'd like it (and its minor releases) to work on 8.9. If CI can only test an old version of the code, then I think there is no value in keeping elpi in the CI for 8.9. If on the contrary the image gets updated I'll happily push to the coq-v8.9 branch each minor fix I do in master (as I do from time to time on coq-master).

**Comment by ejgallego** [comment]
^^ indeed, this was my understanding of the coq-elpi situation for 8.9. Either you drop it or you update it.

**Comment by Zimmi48** [comment]
That's fine with me.

**Comment by silene** [comment]
#8456 has been backported.

**Comment by RalfJung** [comment]
Confirmed, Iris notation tests now pass on the 8.9 branch. Thanks all :)
